File: Lab4/src/Rectangle.java
import java.awt.*;

public class Rectangle implements ISolidShape {

    public Rectangle(
            Color mOutlineColor,
            Color mFillColor,
            Point mLeftTop,
            double mWidth,
            double mHeight
    ) {
    }

    @Override
    public double getArea() {
        return 0;
    }

    @Override
    public double getPerimeter() {
        return 0;
    }

    @Override
    public int getOutlineColor() {
        return 0;
    }

    @Override
    public int getFillColor() {
        return 0;
    }

    @Override
    public String toString() {
        return super.toString();
    }

    public Point getLeftTop() {
        return null;
    }

    public Point getRightBottom() {
        return null;
    }

    public double getWidth() {
        return 0;
    }

    public double getHeight() {
        return 0;
    }
}

File: Lab4/src/shape/Rectangle.kt
package shape

import ICanvas
import ISolidShape
import Point
import java.awt.Color
import java.util.*

class Rectangle(
        private val outlineColor: Color,
        private val fillColor: Color,
        private val leftTop: Point,
        private val width: Double,
        private val height: Double
) : ISolidShape() {

    private val mPoints = Vector<Point>()

    override fun getFillColor() = fillColor
    override fun getOutlineColor() = outlineColor

    override fun getArea(): Double {
        return width * height
    }

    override fun getPerimeter(): Double {
        return 2 * (width + height)
    }

    override fun draw(canvas: ICanvas) {
        if (mPoints.isEmpty()) {
            val rightTop = Point(leftTop.x + width, leftTop.y)
            val leftBottom = Point(leftTop.x, leftTop.y - height)
            val rightBottom = Point(leftTop.x + width, leftTop.y - height)
            mPoints.add(leftTop)
            mPoints.add(rightTop)
            mPoints.add(rightBottom)
            mPoints.add(leftBottom)
        }
        canvas.fillPolygon(mPoints, fillColor)
        canvas.drawPolygon(mPoints, outlineColor)
    }

    override fun stringify(): String {
        return "Rectangle (leftTop=(${leftTop.x},${leftTop.y}), width=$width, height=$height)"
    }
}
